File: my_first_ros_rclcpp_pkg/src/helloworld_subscriber.cpp
#include <functional>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

// placeholders 클래스는 bind 함수의 대체자 역할을 위하여 _1로 선언하였다.
using std::placeholders::_1;

class HelloworldSubscriber : public rclcpp::Node
{
public:
  HelloworldSubscriber()
  : Node("Helloworld_subscriber")
  {
    auto qos_profile = rclcpp::QoS(rclcpp::KeepLast(10));
    // 그 다음으로는 Node 클래스의 create_subscription 함수를 이용해 서브스크라이버를 설정하고 있다.
    // 매개변수의 토픽 메시지 타입은 String, 토픽 이름은 helloworld, QoS 설정은 qos_profile, 콜백 함수는 subscribe_topic_message로 설정하였다.
    helloworld_subscriber_ = this->create_subscription<std_msgs::msg::String>(
      "helloworld",
      qos_profile,
      std::bind(&HelloworldSubscriber::subscribe_topic_message, this, _1));
  }

private:
  // 다음은 앞에서 지정한 subscribe_topic_message 콜백 함수이다.
  // 서브스크라이브한 메시지는 String 타입이고 msg 인자를 통해 받은 메시지는 msg.data 변수에 저장된다.
  void subscribe_topic_message(const std_msgs::msg::String::SharedPtr msg) const
  {
    RCLCPP_INFO(this->get_logger(), "Received message: '%s'", msg->data.c_str());
  }
  // 함수 바깥에는 클래스에서 private 변수로 사용되는 helloworld_subscriber_를 선언하였다.
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr helloworld_subscriber_;
};


int main(int argc, char * argv[])
{
  rclcpp::init(argc,argv);
  auto node = std::make_shared<HelloworldSubscriber>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: my_first_ros_rclpy_pkg/my_first_ros_rclpy_pkg/helloworld subscriber.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile
from std_msgs.msg import String

# 이 노드의 메인 클래스는 HelloworldSubscriber로 Node 클래스를 상속해 사용할 예정이다.
class HelloworldSubscriber(Node):

    def __init__(self):
        # 다음은 클래스 생성자의 정의로 super().__init__('Helloworld_subscriber)를 이용해
        # 부모 클래스(Node)의 생성자를 호출하고 노드 이름을 Helloworld_subscriber로 지정하였다.
        super().__init__('Helloworld_subscriber')
        qos_profile = QoSProfile(depth=10)
        self.helloworld_subscriber = self.create_subscription(
            String,
            'helloworld',
            self.subscribe_topic_message,
            qos_profile)

    # 다음은 앞에서 지정한 subscribe_topic_message 콜백 함수이다.
    # 서브스크라이브한 메시지는 String 타입이며, 실제 데이터는 msg.data 변수에 저장되어 있다.
    def subscribe_topic_message(self, msg):
        self.get_logger().info('Received message: {0}'.format(msg.data))


def main(args=None):
    rclpy.init(args=args)
    node = HelloworldSubscriber()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info('Keyboard Interrupt (SIGINT)')
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
